R/loadFromMetaphlan.R

```diff
@@ -150,18 +150,20 @@ loadFromMetaphlan <- function(file, sample_meta = NULL, phy_tree = NULL, ...){
 
 # Check that metaphlan file contains correct information
 .check_metaphlan <- function(data){
+    # Getting column indices for rowdata and assay
+    col_idx <- unlist(lapply(data, is.character))
     # Get rowdata columns
-    rowdata_columns <- data[ , 1:2]
+    rowdata_columns <- data[ , col_idx, drop=FALSE]
     # Get columns that go to assay
-    assay_columns <- data[ , 3:ncol(data)]
+    assay_columns <- data[ , !col_idx]
     # Initialize result 
     result <- TRUE
     
     # Check rowdata column names that they contain right information, and check that 
     # rest of the columns represents abundances in samples.
     # If these requirements are met, give FALSE. Otherwise, give TRUE.
     if( any(colnames(rowdata_columns) %in% "clade_name") && 
-        any(grepl("id", colnames(rowdata_columns))) && 
+        #any(grepl("id", colnames(rowdata_columns))) && 
         is.numeric(unlist(assay_columns)) ){
```

Comment:
doesn't is.numeric apply directly on assay_columns?

Reply:
apparently not:
![image](https://user-images.githubusercontent.com/86964562/191008816-28329137-f79a-475e-8b1a-69e76c13dafd.png)

Reply:
OK. 

By your definition in the code, the assay_columns seems to be "all non-character columns". 

Runnning unlist() on the matrix is potentially slow. 

Would it be more neat to do a similar check than with `col_idx` definition, something like `col_numeric` and use that to build the if() check instead? This way you would not need to unlist the whole matrix.

Reply:
Is this correct? I have redone it this way:
`# Check that metaphlan file contains correct information
.check_metaphlan <- function(x){
    # Getting column indices for rowdata and assay
    col_idx <- sapply(x, is.character)
    # Get rowdata column indices
    col_rowdata_idx <- col_idx[col_idx==TRUE] 
    # Get assay column indices
    col_assay_idx <- !col_idx[col_idx==FALSE]
    # Initialize result 
    result <- TRUE
    
    # Check rowdata exist
    # Check assay abundance data exist
    # Check rowdata column names that they contain right information.
    # If these requirements are met, give FALSE. Otherwise, give TRUE.
    if( sum(col_rowdata_idx)!=0 && 
        sum(col_assay_idx)!=0 &&
        any(names(col_rowdata_idx) %in% "clade_name") ){
        result <- FALSE
    }
    return(result)
}`

Reply:
Should we accept other than non-numeric assay data? (perhaps yes, I am just asking).

Because if not, then you could check: 1) char cols; 2) numeric cols; 3) other cols.  And generate error if there are any entries in (3). Otherwise, it would be safe to proceed with just picking the character and numeric cols separately and proceeding with them without need for further numerics checks.

Reply:
Sure, no other than numeric assay data should be accepted.
I do agree there could be chances of wrong output from metaphlan or parsing from read.table. It would be much safer as you said to have the three steps checked. Would these steps be good?

```
# Check that metaphlan file contains correct information
.check_metaphlan <- function(x){
    # Getting column indices of character columns
    col_idx <- sapply(x, is.character)
    # Get rowdata column indices
    col_rowdata_idx <- col_idx[col_idx==TRUE]
    # Getting column indices of numeric columns
    col_idx <- sapply(x[,!col_idx], is.numeric)
    # Get assay column indices
    col_assay_idx <- col_idx[col_idx==TRUE]
    # Initialize result 
    result <- TRUE
    
    # Check presence of other data than numeric and character
    # Check rowdata exist
    # Check assay abundance data exist
    # Check rowdata column names that they contain right information.
    # If these requirements are met, give FALSE. Otherwise, give TRUE.
    if( sum(col_idx==FALSE)==0 && 
        sum(col_rowdata_idx)!=0 && 
        sum(col_assay_idx)!=0 &&
        any(names(col_rowdata_idx) %in% "clade_name") ){
        result <- FALSE
    }
    return(result)
}
```